Source/ARTRealtime.m

```diff
@@ -699,42 +701,53 @@ - (ARTEventListener *)transitionSideEffects:(ARTConnectionStateChange *)stateCha
+    for (ARTRealtimeChannelInternal *channel in self.channels.nosyncIterable) {
+        [channels addObject:channel];
+    }
+
     if ([self shouldSendEvents]) {
         [self sendQueuedMessages];
-        // For every Channel
-        for (ARTRealtimeChannelInternal *channel in self.channels.nosyncIterable) {
-            [channel sendQueuedMessages];
+
+        // Channels
+        for (ARTRealtimeChannelInternal *channel in channels) {
+            if (stateChange.previous == ARTRealtimeInitialized ||
+                stateChange.previous == ARTRealtimeConnecting ||
+                stateChange.previous == ARTRealtimeDisconnected) {
+                // RTL4i
+                [channel _attach:nil];
+            }
         }
     } else if (![self shouldQueueEvents]) {
         ARTStatus *channelStatus = status;
         if (!channelStatus) {
             channelStatus = stateChange.reason ? [ARTStatus state:ARTStateError info:stateChange.reason] : [self defaultError];
         }
         [self failQueuedMessages:channelStatus];
-        
-        // If there's a channels.release() going on waiting on this channel
-        // to detach, doing those operations on it here would fire its event listener and
-        // immediately remove the channel from the channels dictionary, thus
-        // invalidating the iterator and causing a crashing.
-        //
-        // So copy the channels and operate on them later, when we're done using the iterator.
-        NSMutableArray<ARTRealtimeChannelInternal *> * const channels = [[NSMutableArray alloc] init];
-        for (ARTRealtimeChannelInternal *channel in self.channels.nosyncIterable) {
-            [channels addObject:channel];
-        }
-        
+
+        // Channels
         for (ARTRealtimeChannelInternal *channel in channels) {
-            if (stateChange.current == ARTRealtimeClosing) {
-                //do nothing. Closed state is coming.
-            }
-            else if (stateChange.current == ARTRealtimeClosed) {
-                [channel detachChannel:[ARTStatus state:ARTStateOk]];
-            }
-            else if (stateChange.current == ARTRealtimeSuspended) {
-                [channel setSuspended:channelStatus];
-            }
-            else {
-                [channel setFailed:channelStatus];
+            switch (stateChange.current) {
+                case ARTRealtimeClosing:
+                    //do nothing. Closed state is coming.
+                    break;
+                case ARTRealtimeClosed:
+                    [channel detachChannel:[ARTStatus state:ARTStateOk]];
+                    break;
+                case ARTRealtimeSuspended:
+                    [channel setSuspended:channelStatus];
+                    break;
+                case ARTRealtimeFailed:
+                    [channel setFailed:channelStatus];
+                    break;
+                default:
+                    break;
```

Comment:
Why is all of this still gated under a condition that fails if `options.queueMessages` is false? This is [RTL3](https://sdk.ably.com/builds/ably/specification/main/features/#RTL3), right? That spec item doesn't mention anything about queueMessages, what do they have to do with each other?

Reply:
See https://github.com/ably/ably-cocoa/pull/1941